sympy/core/equation.py

```diff
@@ -0,0 +1,508 @@
+        """
+        Returns the rhs of the equation.
+        """
+        return self.args[1]
+
+    def as_Boolean(self):
+        """
+        Converts the equation to an Equality.
+        """
+        from .relational import Equality
+        return Equality(self.lhs, self.rhs)
+
+    @property
+    def reversed(self):
+        """
+        Swaps the lhs and the rhs.
+        """
+        return Equation(self.rhs, self.lhs, check=False)
+
+    @property
+    def swap(self):
+        """
+        Synonym for `.reversed`
+        """
+        return self.reversed
+
+    def _applytoexpr(self, expr, func, *args, **kwargs):
+        # Applies a function to an expression checking whether there
+        # is a specialized version associated with the particular type of
+        # expression. Errors will be raised if the function cannot be
+        # applied to an expression.
+        funcname = getattr(func, '__name__')
+        if hasattr(expr, funcname):
+            return getattr(expr, funcname)(*args, **kwargs)
+        else:
+            return func(expr, *args, **kwargs)
+
+    def apply(self, func, *args, side='both', **kwargs):
+        """
+        Apply an operation/function/method to the equation returning the
+        resulting equation.
+
+        Parameters
+        ==========
+
+        func: object
+            object to apply usually a function
+
+        side: 'both', 'lhs', 'rhs', optional
+            Specifies which side of the equation the operation will be applied
+            to. Default is 'both'.
+
+         """
+        lhs = self.lhs
+        rhs = self.rhs
+        if side in ('both', 'lhs'):
+            lhs = self._applytoexpr(self.lhs, func, *args, **kwargs)
+        if side in ('both', 'rhs'):
+            rhs = self._applytoexpr(self.rhs, func, *args, **kwargs)
+        return Equation(lhs, rhs, check=False)
+
+    def applylhs(self, func, *args, **kwargs):
+        """
+        If lhs side of the equation has a defined subfunction (attribute) of
+        name ``func``, that will be applied instead of the global function.
+        The operation is applied to only the lhs.
+        """
+        return self.apply(func, *args, **kwargs, side='lhs')
+
+    def applyrhs(self, func, *args, **kwargs):
+        """
+        If rhs side of the equation has a defined subfunction (attribute) of
+        name ``func``, that will be applied instead of the global function.
+        The operation is applied to only the rhs.
+        """
+        return self.apply(func, *args, **kwargs, side='rhs')
+
+    class _Sides:
+        def __init__(self,eqn, side='both'):
+            self.eqn = eqn
+            self.side = side
+
+        def __getattr__(self, name):
+            func = None
+            try:
+                func = getattr(self.eqn.rhs, name)
+            except AttributeError:
+                func = getattr(self.eqn.lhs, name)
+            return functools.partial(self.eqn.apply, func, side=self.side)
```

Comment:
I would prefer to avoid catching attribute error as it has been used badly in the codebase in the past. You can use:
```
func = getattr(self.eqn.rhs, name, None)
if func is None:
    ...
```
When is it useful to catch attribute error from one side and then use the other side?

Reply:
Was trying to minimize the amount of logic because this may be applied to `self.lhs`, `self.rhs` or both sides. Have changed to:
```
        def __getattr__(self, name):
            func = None
            if self.side in ('rhs', 'both'):
                func = getattr(self.eqn.rhs, name, None)
            else:
                func = getattr(self.eqn.lhs, name, None)
            return functools.partial(self.eqn.apply, func, side=self.side)
```
commit e739037d